### PR TITLE
Add PostClient.saveBatch variant with SQLConnection RMB-354

### PR DESCRIPTION
**Pull request by adamdickmeiss**
Sorry.. some formatting also part of this..

### CODE REVIEW

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java [hunk 1]
```diff
@@ -885,50 +885,55 @@ public void saveBatch(String table, JsonArray entities, Handler<AsyncResult<Resu
...
+        return;
+      }
+      saveBatch(res, table, entities, column, x -> {
+        res.result().close();
+        replyHandler.handle(x);
+      });
+    });
+  }
+
+  public void saveBatch(AsyncResult<SQLConnection> sqlConnection, String table, JsonArray entities, String column, Handler<AsyncResult<ResultSet>> replyHandler) {
```
Comment (OleksiiMaksymov) [inline domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: Plz add another wrapper public method, it is more convenient for work with List<Object>.
as:
`public void saveBatch(AsyncResult<SQLConnection> sqlConnection, String table, List<Object> entities, Handler<AsyncResult<ResultSet>> replyHandler) {....}`

In most cases, we operate on entities, and I think it is more convenient to have two methods with List of Objects and an JsonArray

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java [hunk 2]
```diff
@@ -885,50 +885,55 @@ public void saveBatch(String table, JsonArray entities, Handler<AsyncResult<Resu
...
-    client.getConnection(res -> {
-      if (res.failed()) {
-        replyHandler.handle(Future.failedFuture(res.cause()));
+    if (log.isInfoEnabled()) {
+      log.info("starting: saveBatch size=" + entities.size());
+    }
+    StringBuilder sql = new StringBuilder()
+      .append(INSERT_CLAUSE)
+      .append(schemaName).append(DOT).append(table)
+      .append(" (").append(column).append(") VALUES (?)");
```
Comment (OleksiiMaksymov) [inline domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: ` String column` - I assume that this parameter is redundant to save objects in batch

Comment (adamdickmeiss) [reply domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: Redundant in the "inner" function?? Are you saying that the inner function do not need a column OR that  you just don't need it for YOUR purpose.. Just like the previous one.. WOuld like a version that takes List<Object> and no column parameter?

Comment (OleksiiMaksymov) [reply domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: In our case, the `column `parameter is not needed but may be needed for the future. Currently, we are using in the `mod-calendar` a list of objects without a  `column` parameter

**Comment by adamdickmeiss** [comment]
Note.. I need a test for this.. But want to be sure, this kind of method is what you prefer.

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java [hunk 3]
```diff
@@ -884,63 +884,66 @@ public void saveBatch(String table, JsonArray entities, Handler<AsyncResult<Resu
...
+      if (res.failed()) {
+        replyHandler.handle(Future.failedFuture(res.cause()));
+        return;
+      }
+      saveBatch(res, table, entities, column, x -> {
+        res.result().close();
+        replyHandler.handle(x);
+      });
+    });
+  }
```
Comment (julianladisch) [inline domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: a) This method consists mostly of duplicated code.
b) The res.failed() clause is not tested.

Please use the one liner we use for all new code:

        client.getConnection(conn -> saveBatch(conn, table, entities, column, closeAndHandleResult(conn, replyHandler)));

Move handling the res.failed() code into the saveBatch(res, ...) method.

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java [hunk 4]
```diff
@@ -949,9 +952,35 @@ public void saveBatch(String table, List<Object> entities, Handler<AsyncResult<R
...
+   * @param replyHandler result, containing the id field for each inserted POJO
+   */
+  public void saveBatch(String table, List<Object> entities,
+    Handler<AsyncResult<ResultSet>> replyHandler) {
+
+    JsonArray jsonArray = batchEntitiesToJsonArray(entities, replyHandler);
+    if (jsonArray != null) {
+      saveBatch(table, jsonArray, replyHandler);
+    }
+  }
```
Comment (julianladisch) [inline domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: This is duplicated code of the following saveBatch method.

Please use the usual one-liner to invoke the other saveBatch method instead:

    client.getConnection(conn -> saveBatch(conn, entities, closeAndHandleResult(conn, replyHandler)));

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java [hunk 5]
```diff
@@ -949,9 +952,35 @@ public void saveBatch(String table, List<Object> entities, Handler<AsyncResult<R
...
+
+  public void saveBatch(AsyncResult<SQLConnection> sqlConnection, String table,
+    List<Object> entities, Handler<AsyncResult<ResultSet>> replyHandler) {
+
+    JsonArray jsonArray = batchEntitiesToJsonArray(entities, replyHandler);
+    if (jsonArray != null) {
+      saveBatch(sqlConnection, table, jsonArray, DEFAULT_JSONB_FIELD_NAME, replyHandler);
     }
-    saveBatch(table, jsonArray, replyHandler);
   }
```
Comment (julianladisch) [inline domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: Please add a unit test for jsonArray == null. After that please resubmit for code review!

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java [hunk 6]
```diff
@@ -884,63 +884,66 @@ public void saveBatch(String table, JsonArray entities, Handler<AsyncResult<Resu
...
-  /***
-   * Save a list of POJOs.
-   * POJOs are converted to json and saved in a single sql call. The generated IDs of the
-   * inserted records are returned in the ResultSet.
-   * @param table  destination table to insert into
-   * @param entities  each list element is a POJO
-   * @param replyHandler  result, containing the id field for each inserted POJO
-   */
-  public void saveBatch(String table, List<Object> entities, Handler<AsyncResult<ResultSet>> replyHandler) {
+  private JsonArray batchEntitiesToJsonArray(List<Object> entities,
```
Comment (julianladisch) [inline domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: There is no docum,entation when batchEntitiesToJsonArray invokes the replyHandler.
I don't think that using the replyHandler in this method is a good idea. Simply throw the exception and let the caller handle it, that is more easy to understand and the Exception must be created anyhow.

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java (hunk 5 above)
Comment (julianladisch) [inline domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: For these RMB methods with callbacks we want to make sure that all Exceptions are caught and reported via the replyHandler.
Please wrap into a try-catch clause.

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java [hunk 7]
```diff
@@ -884,63 +884,66 @@ public void saveBatch(String table, JsonArray entities, Handler<AsyncResult<Resu
...
+        return;
+      }
+      saveBatch(res, table, entities, column, x -> {
+        res.result().close();
+        replyHandler.handle(x);
+      });
+    });
+  }
+
+  private void saveBatch(AsyncResult<SQLConnection> sqlConnection, String table,
```
Comment (julianladisch) [inline domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: This should be public.
The corresponding saveBatch method without sqlConnection is already public.

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java [hunk 8]
```diff
@@ -884,63 +884,66 @@ public void saveBatch(String table, JsonArray entities, Handler<AsyncResult<Resu
...
+    StringBuilder sql = new StringBuilder()
+      .append(INSERT_CLAUSE)
+      .append(schemaName).append(DOT).append(table)
+      .append(" (").append(column).append(") VALUES (?)");
+    for (int i = 1; i < entities.size(); i++) {
+      sql.append(",(?)");
+    }
+    sql.append(" RETURNING ").append(idField);
+
+    SQLConnection connection = sqlConnection.result();
```
Comment (julianladisch) [inline domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: Defensive programming: replyHandler needs to be called with a failedFuture if  sqlConnection==null or sqlConnection.failed().

File: domain-models-runtime/src/test/java/org/folio/rest/persist/PostgresClientIT.java [hunk 9]
```diff
@@ -507,6 +509,22 @@ public void saveBatchX(TestContext context) {
...
+  @Test
+  public void saveBatchXTrans(TestContext context) {
+    List<Object> list = Collections.singletonList(xPojo);
+    postgresClient = createFoo(context);
+    postgresClient.startTx(asyncAssertTx(context, trans -> {
+      postgresClient.saveBatch(trans, FOO, list, context.asyncAssertSuccess(save -> {
+        final String id = save.getResults().get(0).getString(0);
+          postgresClient.endTx(trans, context.asyncAssertSuccess(end -> {
+            postgresClient.getById(FOO, id, context.asyncAssertSuccess(get -> {
+            context.assertEquals("x", get.getString("key"));
```
Comment (julianladisch) [inline domain-models-runtime/src/test/java/org/folio/rest/persist/PostgresClientIT.java]: fix indentation

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java (hunk 6 above)
Comment (julianladisch) [inline domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: Use generics for the List type.

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java [hunk 10]
```diff
@@ -949,9 +952,35 @@ public void saveBatch(String table, List<Object> entities, Handler<AsyncResult<R
...
+
+  /***
+   * Save a list of POJOs.
+   * POJOs are converted to json and saved in a single sql call. The generated IDs of the
+   * inserted records are returned in the ResultSet.
+   * @param table  destination table to insert into
+   * @param entities  each list element is a POJO
+   * @param replyHandler result, containing the id field for each inserted POJO
+   */
+  public void saveBatch(String table, List<Object> entities,
```
Comment (julianladisch) [inline domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: Use generics for the List type.

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java [hunk 11]
```diff
@@ -949,9 +952,35 @@ public void saveBatch(String table, List<Object> entities, Handler<AsyncResult<R
...
+    Handler<AsyncResult<ResultSet>> replyHandler) {
+
+    JsonArray jsonArray = batchEntitiesToJsonArray(entities, replyHandler);
+    if (jsonArray != null) {
+      saveBatch(table, jsonArray, replyHandler);
+    }
+  }
+
+  public void saveBatch(AsyncResult<SQLConnection> sqlConnection, String table,
+    List<Object> entities, Handler<AsyncResult<ResultSet>> replyHandler) {
```
Comment (julianladisch) [inline domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: Use generics for the List type.

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java [hunk 12]
```diff
@@ -949,9 +952,35 @@ public void saveBatch(String table, List<Object> entities, Handler<AsyncResult<R
       }
     } catch (Exception e) {
       replyHandler.handle(Future.failedFuture(e));
-      return;
+      return null;
```
Comment (julianladisch) [inline domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: Unit test for null missing.

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java [hunk 13]
```diff
@@ -949,9 +952,35 @@ public void saveBatch(String table, List<Object> entities, Handler<AsyncResult<R
...
+   * @param entities  each list element is a POJO
+   * @param replyHandler result, containing the id field for each inserted POJO
+   */
+  public void saveBatch(String table, List<Object> entities,
+    Handler<AsyncResult<ResultSet>> replyHandler) {
+
+    JsonArray jsonArray = batchEntitiesToJsonArray(entities, replyHandler);
+    if (jsonArray != null) {
+      saveBatch(table, jsonArray, replyHandler);
+    }
```
Comment (julianladisch) [inline domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: Please add unit test for jsonArray == null.
